packages/x-tree-view/src/internals/utils/models.ts

```diff
@@ -0,0 +1,43 @@
+import { warnOnce } from '@mui/x-internals/warning';
+import * as React from 'react';
+
+function useAssertModelConsistencyOutsideOfProduction<T>(parameters: {
+  state: string;
+  controlled: T | undefined;
+  defaultValue: T;
+}) {
+  const { state, controlled, defaultValue } = parameters;
+  const [{ initialDefaultValue, isControlled }] = React.useState({
+    initialDefaultValue: defaultValue,
+    isControlled: controlled !== undefined,
+  });
+
+  if (isControlled !== (controlled !== undefined)) {
+    warnOnce(
+      [
+        `MUI X: A component is changing the ${
+          isControlled ? '' : 'un'
+        }controlled ${state} state of Tree View to be ${isControlled ? 'un' : ''}controlled.`,
```

Comment:
If you remove `Tree view` form the message we could use them too in charts

Reply:
I would add a `componentName` param
Having the name of the component in the error makes the debugging a lot easier imo